add leniency to api v2 conformance statement (#2431)
microsoft/dicom-server: docs/resources/v2-conformance-statement.md

- `multipart/related; type="application/dicom"`
- `application/dicom`

> > Note: The Server will <u>not</u> coerce or replace existing data on the DICOM binary. The DICOM binary will be stored as provided.

Reviewer: Looking through the standard, `DICOM file` is probably the best term to use - and adding the `PS 3.10` part makes it extra clear. https://dicom.nema.org/medical/dicom/current/output/html/part18.html#sect_10.5.1.4

```suggestion
> Note: the Server will <u>not</u> coerce or replace data in the DICOM PS 3.10 file. The DICOM file will be stored as provided, except where otherwise noted.
```

Each file stored must have a unique combination of StudyInstanceUID, SeriesInstanceUID and SopInstanceUID. The warning code `45070` will be returned if a file with the same identifiers already exists.

> > DICOM File Size Limit: there is a size limit of 2GB for request size, thus limiting the size of the DICOM file.

Reviewer: ```suggestion
> Requests are limited to 2GB. No single DICOM file or combination of files may exceed this limit.
```

### Store Response Status Codes

| Code                         | Description                                                                                                                                                                                                                         |
| :--------------------------- |:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 200 (OK)                     | All the SOP instances in the request have been stored.                                                                                                                                                                              |

> | 202 (Accepted)               | The origin server stored some of the Instances and failures exist for others. Or origin server has warnings for Instances stored. Additional information regarding this error may be found in the response message body. |

Reviewer 2: I think we can tighten this up a little: 

```suggestion
| 202 (Accepted)               | The origin server stored some of the Instances and others have failed or returned warnings. Additional information regarding this error may be found in the response message body. |
```

Each file stored must have a unique combination of StudyInstanceUID, SeriesInstanceUID and SopInstanceUID. The warning code `45070` will be returned if a file with the same identifiers already exists.

> Requests are limited to 2GB. No single DICOM file or combination of files may exceed this limit.

> ### Store Changes From V1

Reviewer 2: Good addition 😄 Provides a lot of clarity about the changes.

> A major change from V1 Store API is that attributes will continue to be validated, but only validation errors with [required attributes](#store-required-attributes) would cause for Store to fail.

Reviewer 2: Minor suggestion: 

```suggestion
A major change from V1 Store API is that attributes will continue to be validated, but only validation errors with [required attributes](#store-required-attributes) would cause Store to fail.
```

Failed validation of attributes not required by the API will still result in the file being stored and a warning will be given about each failing attribute per instance.

> In the case of a failing sequence or multiple issues of a single attribute, only the first failing attribute reason will be noted.

Reviewer: ```suggestion
When a sequence contains an attribute that fails validation, or when there are multiple issues with a single attribute, only the first failing attribute reason will be noted.
```

      {
        "vr":"UI",
        "Value":["d246deb5-18c8-4336-a591-aeb6f8596664"]
      },
      "00081155":
      {
        "vr":"UI",
        "Value":["4a858cbb-a71f-4c01-b9b5-85f88b031365"]
      },
      "00081190":
      {
        "vr":"UR",
        "Value":["http://localhost/studies/d09e8215-e1e1-4c7a-8496-b4f6641ed232/series/8c4915f5-cc54-4e50-aa1f-9b06f6e58485/instances/4a858cbb-a71f-4c01-b9b5-85f88b031365"]
      },
      "00081196": {
        "vr": "US",
        "Value": [
            1
        ]
      },
      "00741048": {
        "vr": "SQ",
        "Value": [
          {
            "00000902": {
              "vr": "LO",
              "Value": [
                "DICOM100: (0008,0020) - Content \"NotAValidDate\" does not validate VR DA: one of the date values does not match the pattern YYYYMMDD"
              ]
            }
          },
          {
            "00000902": {
              "vr": "LO",
              "Value": [
                "DICOM100: (0008,002a) - Content \"NotAValidDate\" does not validate VR DT: value does not mach pattern YYYY[MM[DD[HH[MM[SS[.F{1-6}]]]]]]"
              ]
            }
          }
        ]
      }
    }]
  }
}
```

> In the case that leniency is applied and the file is stored, you will be notified with each failing attribute per instance. In case of a failing sequence or multiple issues of a single attribute, only the first failing attribute reason will be noted.

Reviewer 2: This bit about "leniency is applied" made me question - "what do they mean by leniency?" since it is not mentioned before this in the conformance statement.  

I wonder if we should rephrase this to talk about validation.  Maybe something like:

```suggestion
For attributes that are not required but are searchable by default, validation will be performed.  Failed validation of these attributes will still result in the file being stored and a warning will be given about each failing attribute per instance. In the case of a failing sequence or multiple issues of a single attribute, only the first failing attribute reason will be noted.
```

Reviewer: I think we should have a whole heading discussing leniency on STOW since it's a major change from v1.

Reviewer 2: I agree with Josiah - it would be good to just have a section dedicated to this change and explain the nuances in the behavior.

> We will also output diagnostic logs per failed attribute per instance.

Reviewer 2: I think we should avoid using "we" here.  I know "we" is used in other places, but it would be a good practice start changing.  

```suggestion
Diagnostic logs will also list failed attributes per instance.
```

| 43265 | The provided instance StudyInstanceUID did not match the specified StudyInstanceUID in the store request. |
| 45070 | A DICOM instance with the same StudyInstanceUID, SeriesInstanceUID and SopInstanceUID has already been stored. If you wish to update the contents, delete this instance first. |
| 45071 | A DICOM instance is being created by another process, or the previous attempt to create has failed and the cleanup process has not had chance to clean up yet. Please delete the instance first before attempting to create again. |

### Store Warning Reason Codes

| Code  | Description                                                                                                                                            |
|:------|:-------------------------------------------------------------------------------------------------------------------------------------------------------|
| 45063 | The Studies Store Transaction (Section 10.5) observed that the Data Set did not match the constraints of the SOP Class during storage of the instance. |

> | 1     | The Studies Store Transaction (Section 10.5) observed that the Data Set has validation warnings.                                                       |

Reviewer 2: This having a value of `1` surprised me.  Is that because there is no suitable code defined in the standard?  I see that this page only lists 3: https://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_i.2.html

- `application/dicom+json`

> Note that any instances stored with leniency applied to indexed attributes may not be retrievable by attributes which failed validation.

Reviewer 2: Similar to the above comment, the reader might not know what we mean by "leniency" in this context.  Being more explicit here might be helpful: 

```suggestion
Note if an instance returned validation warnings for searchable attributes at the time the instance was stored, those attributes may not be retrievable via search.
```

- `application/dicom+json`

### Search Changes From V1

> Note if an instance returned validation warnings for [searchable attributes](#searchable-attributes) at the time the [instance was stored](#store-changes-from-v1), those attributes may not be used to search for the stored instance.

Reviewer 2: Minor suggestion:
```suggestion
If an instance returned validation warnings for [searchable attributes](#searchable-attributes) at the time the [instance was stored](#store-changes-from-v1), those attributes may not be used to search for the stored instance.
```
